api/snapshot.go

```diff
@@ -0,0 +1,691 @@
+	if err != nil {
+		return prepareResponseWithErr(http.StatusServiceUnavailable, err)
+	}
+	if isRunning {
+		return prepareResponseWithErr(http.StatusServiceUnavailable, errors.New("Job is already running"))
+	}
+
+	_, err = os.Stat(config.FlagSnapshotDir)
+	if os.IsNotExist(err) {
+		log.Infof("snapshot dir: %s not found, attempting to create one", config.FlagSnapshotDir)
+		if err := os.Mkdir(config.FlagSnapshotDir, os.ModePerm); err != nil {
+			j.Status = "Could not create snapshot directory: " + config.FlagSnapshotDir
+			return prepareResponseWithErr(http.StatusServiceUnavailable, errors.New(j.Status))
+		}
+	}
+
+	// Null errors on every new run.
+	j.Errors = nil
+
+	t := time.Now()
+	j.LastSnapshotPath = fmt.Sprintf("%s/snapshot-%d-%02d-%02dT%02d:%02d:%02d-%d.zip", config.FlagSnapshotDir, t.Year(),
+		t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond())
+	j.Status = "Snapshot job started, archive will be available: " + j.LastSnapshotPath
+
+	// first discover all nodes in a cluster, then try to find requested nodes.
+	masterNodes, err := DCOSTools.GetMasterNodes()
+	if err != nil {
+		return prepareResponseWithErr(http.StatusServiceUnavailable, err)
+	}
+	agentNodes, err := DCOSTools.GetAgentNodes()
+	if err != nil {
+		return prepareResponseWithErr(http.StatusServiceUnavailable, err)
+	}
+	foundNodes, err := findRequestedNodes(masterNodes, agentNodes, req.Nodes)
+	if err != nil {
+		return prepareResponseWithErr(http.StatusServiceUnavailable, err)
+	}
+	log.Debugf("Found requested nodes: %s", foundNodes)
+
+	j.cancelChan = make(chan bool)
+	go j.runBackgroundJob(foundNodes, config, DCOSTools)
+	return prepareResponseOk(http.StatusOK, "Snapshot job started: "+filepath.Base(j.LastSnapshotPath))
+}
+
+//
+func (j *SnapshotJob) runBackgroundJob(nodes []Node, config *Config, DCOSTools DCOSHelper) {
+	log.Info("Started background job")
+	// log a start time
+	j.JobStarted = time.Now()
+
+	// log end time
+	defer func(j *SnapshotJob) {
+		j.JobEnded = time.Now()
+		j.JobDuration = time.Since(j.JobStarted)
+		log.Info("Job finished")
+	}(j)
+
+	// lets start a goroutine which will timeout background report job after a certain time.
+	jobIsDone := make(chan bool)
+	go func(jobIsDone chan bool, j *SnapshotJob) {
+		select {
+		case <-jobIsDone:
+			return
+		case <-time.After(time.Minute * time.Duration(config.FlagSnapshotJobTimeoutMinutes)):
+			j.Status = "Job failed"
+			errMsg := fmt.Sprintf("snapshot job timedout after: %s", time.Since(j.JobStarted))
+			j.Errors = append(j.Errors, errMsg)
+			log.Error(errMsg)
+			j.cancelChan <- true
+			return
+		}
+	}(jobIsDone, j)
+
+	// makesure we always cancel a timeout goroutine when the report is finished.
+	defer func(jobIsDone chan bool) {
+		jobIsDone <- true
+	}(jobIsDone)
+
+	// Update job running field.
+	j.start()
+	defer j.stop()
+
+	// create a zip file
+	zipfile, err := os.Create(j.LastSnapshotPath)
+	if err != nil {
+		j.Status = "Job failed"
```

Comment:
It seems like `.Status` and `.Errors` both provide stringified status messages but are somewhat duplicated results. You could make `.Status` a bool and only be verbose with `.Errors` when `.Status` is false (I guess status could be `.Success` in this context).